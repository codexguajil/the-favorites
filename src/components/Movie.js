import React, { Component } from 'react';

export class Movie extends Component {
  render() {
    const { title, rating, posterImage } = this.props
    return (
      <div>
        <h4>{title}</h4>
        <p>{rating}</p>
        <img src={posterImage} alt='movie poster'/>
      </div>
    )
  }
}

export default Movie
