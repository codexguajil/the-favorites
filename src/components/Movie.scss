@font-face {
  font-family: 'movie-title';
  src: url('./utils/assets/neue-helv/HelveticaNeue-LightCondObl.woff') format('woff2');
}

.movie {
  width: 200px;
  margin: 15px;
  padding-top: 40px;
  position: relative;
  img {
    width: 200px;
  }
  .movie-info {
    background: hsla(0, 0%, 5%, 1);
    color: white;
    visibility: hidden;
    font-family: 'arial';
    height: 200px;
    width: 400px;
    position: fixed;
    // top: 250px;
    // margin-left: 100px;
    // // overflow-y: auto;
    text-overflow: ellipsis;
    z-index: 1;
  }
  &:hover {
    .movie-info {
      visibility: visible;
      // position: fixed;
      // padding: 10px;
      // transform: none;
      // right: auto;
      // bottom: auto;
      // top: auto;
      // left: auto;
      // margin: 0 auto;
      // margin-top: -15px;
      // margin-left: -22px;
    }
  &:after {
    .movie-info {
    }
  }
  }
  // &:after {
  //     border-left: 20px solid transparent;
  //     border-top: 20px solid #efefef;
  //     bottom: -20px;
  //     content: "";
  //     position: absolute;
  //     right: 20px;
  // }
}

.movie-title {
  color: 'black';;
  font-size: 22px;
  padding: 15px;
}

a {
  text-decoration: none;
  color: 'white';
}