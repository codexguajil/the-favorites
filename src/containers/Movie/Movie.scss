@font-face {
  font-family: 'movie-title';
  src: url('./utils/assets/neue-helv/HelveticaNeue-LightCondObl.woff') format('woff2');
}

.movie {
  width: 200px;
  margin: 15px;
  padding-top: 40px;
  position: relative;
  img {
    opacity: .75;
    width: 200px;
  }
  .movie-info {
    background: hsla(0, 0%, 5%, 1);
    color: white;
    visibility: hidden;
    font-family: 'arial';
    height: 200px;
    width: 400px;
    position: fixed;
    text-overflow: ellipsis;
    z-index: 1;
  }
  &:hover {
    img {
      opacity: 1;
      box-shadow: 1px 1px 12px 3px hsla(0, 0%, 5%, 1);
    }
  }
}

.fav-false {
  height: 65px;
  width: 65px;
  position: absolute;
  top: 24px;
  left: -19px;
  background: url('./utils/assets/false.png');
  background-repeat: no-repeat;
}

.fav-true {
  height: 65px;
  width: 65px;
  position: absolute;
  top: 24px;
  left: -19px;
  background: url('./utils/assets/true.png')
}

.movie-title {
  color: 'black';;
  font-size: 22px;
  padding: 15px;
}

a {
  text-decoration: none;
  color: 'white';
}