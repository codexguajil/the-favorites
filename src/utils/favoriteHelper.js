export const  validateFavorite = async (movie) => {
    const { user_id, id } = this.props
    const url = `http://localhost:3000/api/users/${user_id}/favorites`      
    if(user_id) {
      const allFavorites = await fetchData(url)
      allFavorites.data.forEach(favorite => {
        if(favorite.movie_id === id) {
          console.log('alreay in favorites')
        } else {
          this.addFavorite()
        }
      })
    } else {
      this.setState()
    }
  }

export const  addFavorite = async (e) => {
    const { id, title, rating, user_id, posterImage, synopsis, releaseDate } = this.props
    const url = "http://localhost:3000/api/users/favorites/new"
    const movie = {
      movie_id: id, 
      user_id: user_id, 
      title: title, 
      poster_path: posterImage, 
      release_date: releaseDate, 
      vote_average: rating, 
      overview: synopsis
    }
    try {
      await fetchData(url, {
        method: "POST", 
        body: JSON.stringify(movie),
        headers: {
          "Content-Type": "application/json"
        }
      })
    } catch(error) {
      console.log(error.message)
    }
  }