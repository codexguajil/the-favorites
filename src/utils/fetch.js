export const fetchData = async (url, options) => {
  const response = await fetch(url, options)
  console.log(response)
  if(response.ok) {
    const parsedData = response.json()
    return parsedData;
  } else {
    return response
  }
}