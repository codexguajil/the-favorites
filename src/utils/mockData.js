export const mockUser = {
  id: 1,
  name: "Taylor",
  password: "password",
  email: "tman2272@aol.com"
}

export const mockMovieResponse = {
page: 1,
results: [
  {adult: false,
  backdrop_path: "/hMANgfPHR1tRObNp2oPiOi9mMlz.jpg",
  genre_ids: [28, 12, 878],
  id: 424783,
  original_language: "en",
  original_title: "Bumblebee",
  overview: "On the run in the year 1987, Bumblebee finds refuge in a junkyard in a small Californian beach town. Charlie, on the cusp of turning 18 and trying to find her place in the world, discovers Bumblebee, battle-scarred and broken.  When Charlie revives him, she quickly learns this is no ordinary yellow VW bug.",
  popularity: 93.533,
  poster_path: "/fw02ONlDhrYjTSZV8XO6hhU3ds3.jpg",
  release_date: "2018-12-15",
  title: "Bumblebee",
  video: false,
  vote_average: 6.5,
  vote_count: 1414},
  {adult: false,
  backdrop_path: "/9QusGjxcYvfPD1THg6oW3RLeNn7.jpg",
  genre_ids:[28, 12, 14, 878],
  id: 297802,
  original_language: "en",
  original_title: "Aquaman",
  overview: "Once home to the most advanced civilization on Earth, Atlantis is now an underwater kingdom ruled by the power-hungry King Orm. With a vast army at his disposal, Orm plans to conquer the remaining oceanic people and then the surface world. Standing in his way is Arthur Curry, Orm's half-human, half-Atlantean brother and true heir to the throne.",
  popularity: 165.687,
  poster_path: "/5Kg76ldv7VxeX9YlcQXiowHgdX6.jpg",
  release_date: "2018-12-07",
  title: "Aquaman",
  video: false,
  vote_average: 6.8,
  vote_count: 4652,
  },
  {adult: false,
  backdrop_path: "/bTeRgkAavyw1eCtSkaww18wLYNP.jpg",
  genre_ids:[80, 18, 53],
  id: 504172,
  original_language: "en",
  original_title: "The Mule",
  overview: "Earl Stone, a man in his 80s who is broke, alone, and facing foreclosure of his business when he is offered a job that simply requires him to drive. Easy enough, but, unbeknownst to Earl, he’s just signed on as a drug courier for a Mexican cartel. He does so well that his cargo increases exponentially, and Earl hit the radar of hard-charging DEA agent Colin Bates.",
  popularity: 128.205,
  poster_path: "/oeZh7yEz3PMnZLgBPhrafFHRbVz.jpg",
  release_date: "2018-12-14",
  title: "The Mule",
  video: false,
  vote_average: 6.5,
  vote_count: 876},
  {adult: false,
  backdrop_path: "/uUiId6cG32JSRI6RyBQSvQtLjz2.jpg",
  genre_ids:[28, 12, 16, 878, 35],
  id: 324857,
  original_language: "en",
  original_title: "Spider-Man: Into the Spider-Verse",
  overview: "Miles Morales is juggling his life between being a high school student and being a spider-man. When Wilson Kingpin Fisk uses a super collider, others from across the Spider-Verse are transported to this dimension.",
  popularity: 157.433,
  poster_path: "/iiZZdoQBEYBv6id8su7ImL0oCbD.jpg",
  release_date: "2018-12-07",
  title: "Spider-Man: Into the Spider-Verse",
  video: false,
  vote_average: 8.5,
  vote_count: 2715},
  {backdrop_path: "/e7VzDMrYKXVrVon04Uqsrcgnf1k.jpg",
  first_air_date: "2019-03-15",
  genre_ids: [16, 10765],
  id: 86831,
  name: "Love, Death & Robots",
  origin_country: [],
  original_language: "en",
  original_name: "Love, Death & Robots",
  overview: "Terrifying creatures, wicked surprises and dark comedy converge in this NSFW anthology of animated stories presented by Tim Miller and David Fincher.",
  popularity: 161.862,
  poster_path: "/rlrRI2b6RvM9I9xOBTKqcTaehkE.jpg",
  vote_average: 8.6,
  vote_count: 29}],
  total_pages: 1000,
  total_results: 20000
}


export const mockDataResponse = [
  {id: 424783,
  posterImage: "https://image.tmdb.org/t/p/w500/fw02ONlDhrYjTSZV8XO6hhU3ds3.jpg",
  rating: 6.5,
  releaseDate: "2018-12-15",
  synopsis: "On the run in the year 1987, Bumblebee finds refuge in a junkyard in a small Californian beach town. Charlie, on the cusp of turning 18 and trying to find her place in the world, discovers Bumblebee, battle-scarred and broken.  When Charlie revives him, she quickly learns this is no ordinary yellow VW bug.",
  title: "Bumblebee"},
  {id: 297802,
  posterImage: "https://image.tmdb.org/t/p/w500/5Kg76ldv7VxeX9YlcQXiowHgdX6.jpg",
  rating: 6.8,
  releaseDate: "2018-12-07",
  synopsis: "Once home to the most advanced civilization on Earth, Atlantis is now an underwater kingdom ruled by the power-hungry King Orm. With a vast army at his disposal, Orm plans to conquer the remaining oceanic people and then the surface world. Standing in his way is Arthur Curry, Orm's half-human, half-Atlantean brother and true heir to the throne.",
  title: "Aquaman"},
  {id: 504172,
  posterImage: "https://image.tmdb.org/t/p/w500/oeZh7yEz3PMnZLgBPhrafFHRbVz.jpg",
  rating: 6.5,
  releaseDate: "2018-12-14",
  synopsis: "Earl Stone, a man in his 80s who is broke, alone, and facing foreclosure of his business when he is offered a job that simply requires him to drive. Easy enough, but, unbeknownst to Earl, he’s just signed on as a drug courier for a Mexican cartel. He does so well that his cargo increases exponentially, and Earl hit the radar of hard-charging DEA agent Colin Bates.",
  title: "The Mule"},
  {id: 399361,
  posterImage: "https://image.tmdb.org/t/p/w500/aBw8zYuAljVM1FeK5bZKITPH8ZD.jpg",
  rating: 6.3,
  releaseDate: "2019-03-06",
  synopsis: "Struggling to make ends meet, former special ops soldiers reunite for a high-stakes heist: stealing $75 million from a South American drug lord.",
  title: "Triple Frontier"},
  {id: 86831,
  posterImage: "https://image.tmdb.org/t/p/w500/rlrRI2b6RvM9I9xOBTKqcTaehkE.jpg",
  rating: 8.6,
  releaseDate: undefined,
  synopsis: "Terrifying creatures, wicked surprises and dark comedy converge in this NSFW anthology of animated stories presented by Tim Miller and David Fincher.",
  title: "Love, Death & Robots"}
]