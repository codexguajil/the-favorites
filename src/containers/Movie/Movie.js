import React, {Component} from 'react';
import { fetchData } from '../../utils/fetch';
import { addFavorite, deleteFavorite, addMessage } from '../../actions';
import { connect } from 'react-redux';
import { Link } from 'react-router-dom';
import propTypes from 'prop-types';

export class Movie extends Component {
  constructor() {
    super()
    this.state = {
      favorite: false,
      message: ''
    }
  }

  componentDidMount = () => {
    const {favorites, id} = this.props
    favorites.forEach(favorite => {
      if(favorite === id) {
        this.setState({favorite: true})
      }
    })
  }

  validateFavorite = () => {
    const { user_id, id, addMessage } = this.props  
    if(user_id) {
      if(this.props.favorites.includes(id)) {
        this.removeFavorite()
        this.setState({favorite: false})
      } else {
        this.addFavorite()
        this.setState({favorite: true})
      }
    } else {
      addMessage('Please log in to add to favorites.')
      setTimeout(() => {
        addMessage('')
      }, 3000)
    }
  }

  removeFavorite = async () => {
    const { id, user_id } = this.props
    const url = `http://localhost:3000/api/users/${user_id}/favorites/${id}`;
    const userOptionObject = {
      method: "DELETE",
      headers: {
        "Content-Type": "application/json"
      }
    }
    await fetchData(url, userOptionObject);
    this.props.deleteFavorite(id)
  }

  addFavorite = async (e) => {
    const { id, title, rating, user_id, posterImage, synopsis, releaseDate } = this.props
    const url = "http://localhost:3000/api/users/favorites/new"
    const movie = {
      movie_id: id, 
      user_id: user_id, 
      title: title, 
      poster_path: posterImage, 
      release_date: releaseDate, 
      vote_average: rating, 
      overview: synopsis
    }
    try {
      const addedFavorite = await fetchData(url, {
        method: "POST", 
        body: JSON.stringify(movie),
        headers: {
          "Content-Type": "application/json"
        }
      })
      // console.log(addedFavorite)
      this.props.addFavoriteToState(id)
    } catch(error) {
      console.log(error.message)
    }
  }

  render() {
    const {favorite} = this.state
    const { id, title, rating, posterImage, synopsis, type } = this.props
    const image = 'https://image.tmdb.org/t/p/w500'+ posterImage
    return (
      <div className="movie">
        <div className="movie-info">
        {title}
        {rating}
        {synopsis}
        </div>
        {/* <h4 className="movie-title">{title}</h4> */}
        {/* <p classsName="rating">{rating}</p> */}
        <Link to={`/${type}/${id}`}>
          <img src={image} alt='movie poster'/> 
        </Link>
        <div onClick={this.validateFavorite} className={favorite ? 'fav-true': 'fav-false'}></div>
      </div>
    )
  }
}

export const mapStateToProps = (state) => ({
  user_id: state.user.id,
  favorites: state.favorites
})

export const mapDispatchToProps = (dispatch) => ({
  addFavoriteToState: (favorite) => dispatch(addFavorite(favorite)),
  deleteFavorite: (favorite) => dispatch(deleteFavorite(favorite)),
  addMessage: (message) => dispatch(addMessage(message))
})

export default connect(mapStateToProps, mapDispatchToProps)(Movie)

Movie.propTypes = {
  user_id: propTypes.number,
  favorites: propTypes.array.isRequired,
  addFavoriteToState: propTypes.func.isRequired,
  deleteFavorite: propTypes.func.isRequired,
  addMessage: propTypes.func.isRequired
}
