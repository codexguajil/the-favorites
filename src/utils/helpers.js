import { fetchData } from './fetch';

export const cleanMovieData = (movies) => {
  return movies.results.map(movie => {
    return ({
        title: movie.title || movie.name,
        rating: movie.vote_average,
        id: movie.id,
        releaseDate: movie.release_date,
        synopsis: movie.overview,
        posterImage: movie.poster_path
    })
  })
}

export const fetchAllFavorites = async (id) => {
  const url = `http://localhost:3000/api/users/${id}/favorites`;
  const fetchedFavorites = await fetchData(url);
  const cleanedFavorites = cleanFavorites(fetchedFavorites.data)
  return cleanedFavorites;
}

const cleanFavorites = (favorites) => {
  return favorites.map(movie => {
    return movie.movie_id
  })
}