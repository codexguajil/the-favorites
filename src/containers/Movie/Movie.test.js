import React from 'react'
import { Movie, mapDispatchToProps, mapStateToProps } from './Movie'
import { shallow } from 'enzyme'
import { mockMovieResponse } from '../../utils/mockData'


describe('Movie', () => {
  let wrapper;
  let mockFn = jest.fn()

  beforeEach(() => {
    wrapper = shallow(
      <Movie favorites={[1, 2, 3]} 
             deleteFavorite={mockFn} 
             addFavoriteToState={mockFn} 
             user_id={1}
             id={123}
             addMessage={mockFn}
            />
    )

    window.fetch = jest.fn().mockImplementation(() => Promise.resolve({
      ok: true,
      status: 200,
      json: () => Promise.resolve(mockMovieResponse)
    }))
  });

  it('should match snapshot', () => {
    expect(wrapper).toMatchSnapshot()
  });

  describe('removeFavorite', () => {
    it('should delete a favorite from the api', async () => {


      wrapper.instance().removeFavorite()
      expect(window.fetch).toHaveBeenCalled()
    })

    it('should call deleteFavorite', () => {

      wrapper.instance().removeFavorite()
      expect(mockFn).toHaveBeenCalled()
    })

  })

  describe('addFavorite', () => {
    it('should add a favorite', () => {

      wrapper.instance().addFavorite()

      expect(window.fetch).toHaveBeenCalled()
    })

    it('should call addFavoriteToState', () => {

      wrapper.instance().removeFavorite()
      expect(mockFn).toHaveBeenCalled()
    })
  })

  describe('validateFavorite', () => {
    it('should check if a favorite exists', () => {

      wrapper.instance().validateFavorite()

      expect(mockFn).toHaveBeenCalled()
    })

    it('should display a message if there is no user id', () => {

      wrapper = shallow(
        <Movie favorites={[1, 2, 3]} 
               deleteFavorite={mockFn} 
               addFavoriteToState={mockFn} 
               user_id={false}
               id={123}
               addMessage={mockFn}
              />
      )

      wrapper.instance().validateFavorite()

      expect(mockFn).toHaveBeenCalled()
    })
  })

  describe('mapStateToProps', () => {
    it('should show previous value for state favorites', () => {
      const initialState = {
        user: {id: 2},
        favorites: [1, 2, 3]
      }

      expect(mapStateToProps(initialState).favorites).toEqual([1, 2, 3])
    });

    it.skip('should show previous value for state user id', () => {
      const initialState = {
        user: {id: 2},
        favorites: [1, 2, 3]
      }

      expect(mapStateToProps(initialState).user.id).toEqual(2)
    })
  })

  describe('mapDispatchToProps', () => {
    it('should pass correct type when addFavoriteToStore is called', () => {
      const dispatch = jest.fn()

      mapDispatchToProps(dispatch).addFavoriteToState()
      expect(dispatch.mock.calls[0][0]).toEqual({ type: 'ADD_FAVORITE' })
    });

    it('should pass correct type when deleteFavorite is called', () => {
      const dispatch = jest.fn()

      mapDispatchToProps(dispatch).deleteFavorite()
      expect(dispatch.mock.calls[0][0]).toEqual({ type: 'DELETE_FAVORITE' })
    })

    it('should pass correct type when addMessage is called', () => {
      const dispatch = jest.fn()

      mapDispatchToProps(dispatch).addMessage()
      expect(dispatch.mock.calls[0][0]).toEqual({ type: 'ADD_MESSAGE' })
    })
  })
})