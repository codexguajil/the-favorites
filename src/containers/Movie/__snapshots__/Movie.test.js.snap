// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Movie should match snapshot 1`] = `undefined`;
